doc/pandas-2.0/source/internal-architecture.rst

```diff
@@ -0,0 +1,714 @@
+``numpy.dtype`` internally, we can:
+
+* Begin to better protect users from NumPy data semantics (which are frequently
+  different from pandas's!) leaking through to the pandas user API. This can
+  enable us to address long-standing inconsistencies or "rough edges" in pandas
+  that have persisted due to our tight semantic coupling to NumPy.
+
+* We can consider adding new data structures to pandas, either custom to pandas
+  or provided by 3rd-party libraries, that add new functionality alongside the
+  existing code (presuming NumPy physical storage). As one concrete example,
+  discussed in more detail below, we can enable missing data in integer pandas
+  data by forming a composite data structure consisting of a NumPy array plus a
+  bitmap marking the null / not-null values.
+
+* We can start to think about improved behavior around data ownership (like
+  copy-on-write) which may yield many benefits. I will write a dedicated
+  section about this.
+
+Note that neither of these points implies that we are trying to use NumPy
+less. We already have large amounts of code that implement algorithms similar
+to those found in NumPy (e.g. ``pandas.unique`` or the implementation of
+``Series.sum``), but taking into account pandas's missing data representation,
+etc. Internally, we can use NumPy when its computational semantics match those
+we've chosen for pandas, and elsewhere we can invoke pandas-specific code.
+
+A major concern here based on these ideas is **preserving NumPy
+interoperability**, so I'll examine this topic in some detail next.
+
+Preserving NumPy interoperability
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Some of types of intended interoperability between NumPy and pandas are as
+follows:
+
+* Users can obtain the a ``numpy.ndarray`` (possibly a view depending on the
+  internal block structure, more on this soon) in constant time and without
+  copying the actual data. This has a couple other implications
+
+  * Changes made to this array will be reflected in the source pandas object.
+  * If you write C extension code (possibly in Cython) and respect pandas's
+    missing data details, you can invoke certain kinds of fast custom code on
+    pandas data (but it's somewhat inflexible -- see the latest discussion on
+    adding a native code API to pandas).
+
+* NumPy ufuncs (like ``np.sqrt`` or ``np.log``) can be invoked on
+  pandas objects like Series and DataFrame
+
+* ``numpy.asarray`` will always yield some array, even if it discards metadata
+  or has to create a new array. For example ``asarray`` invoked on
+  ``pandas.Categorical`` yields a reconstructed array (rather than either the
+  categories or codes internal arrays)
+
+* Many NumPy methods designed to work on subclasses (or duck-typed classes) of
+  ``ndarray`` may be used. For example ``numpy.sum`` may be used on a Series
+  even though it does not invoke NumPy's internal C sum algorithm. This means
```

Comment:
I deliberately chose pretty hedgy language here. "Many methods ... may be used" instead of "All methods ... can be used". Maybe we should say that we're going to back away from this here?

Maintaining an explicit API contract that you can pass a Series or DataFrame wherever you might otherwise pass an ndarray seems like a bad idea to me, and is likely to be a continuous source of bugs and maintainability problems. I know you've worked a lot of this and very honestly I cringed when I saw the patches come through — I would rather not reinforce this interchangeability in users' minds. I agree that accommodating NumPy functions where it does no harm to pandas is okay, but I don't think it's our responsibility to do so.

But honestly: the average pandas user is, in my anecdotal experience, more familiar with pandas's features than NumPy's features, so the audience of people for whom "semantic parity" with NumPy is important is probably growing smaller over time. I would rather focus on creating a consistent and self-contained user experience in pandas.